src/main/scala/com/ubirch/services/IncidentListener.scala

```diff
@@ -77,49 +78,80 @@ class IncidentListener @Inject()(config: Config, lifecycle: Lifecycle, tenantRet
 
     Future.sequence(consumerRecords.map { cr =>
 
-      try {
-        val hwId = retrieveHeader(cr, HeaderKeys.X_UBIRCH_HARDWARE_ID)
-        val authToken = retrieveHeader(cr, HeaderKeys.X_UBIRCH_DEVICE_INFO_TOKEN)
-        val incident = createIncidentFromCR(cr, hwId)
-
-        tenantRetriever.getDevice(hwId, authToken).map {
-
-          case Some(device: SimpleDeviceInfo) =>
-            logger.info(s"processing incident $incident for owner ${device.customerId} and forwarding it to mqtt")
-            distributor.sendIncident(write(incident).getBytes(StandardCharsets.UTF_8), device.customerId)
+      // it's short circuit evaluation.
+      // if an error occurs in the for comprehension before yield is called, incident is not sent, Left is returned.
+      (for {
+        hwId <- retrieveHeader(cr, HeaderKeys.X_UBIRCH_HARDWARE_ID)
+        authToken <- retrieveHeader(cr, HeaderKeys.X_UBIRCH_DEVICE_INFO_TOKEN)
+        errorCodeOpt <- retrieveHeaderOpt(cr, HeaderKeys.X_CODE)
+        incident <- createIncidentFromCR(cr, hwId, errorCodeOpt)
+      } yield (hwId, authToken, incident)) match {
+        case Right((hwId, authToken, incident)) => {
+          tenantRetriever.getDevice(hwId, authToken).map {
+            case Some(device: SimpleDeviceInfo) =>
+              logger.info(s"processing incident $incident for owner ${device.customerId} and forwarding it to mqtt")
+              distributor.sendIncident(write(incident).getBytes(StandardCharsets.UTF_8), device.customerId)
+            case None =>
```

Comment:
I added this code when a device is not found.
Is it correct logic?

Reply:
Oh, I just didn't treat None? hehe. Yes, that looks good!